igvc_perception/src/pointcloud_filter/fast_segment_filter/fast_segment_filter.cpp

```diff
@@ -0,0 +1,317 @@
+    mean(0, 2) += pt.point_.z;
+  }
+  mean /= -total_points;
+  A += mean.replicate(total_points, 1);
+  Eigen::JacobiSVD<MatrixXd> svd(A, Eigen::ComputeFullU | Eigen::ComputeFullV);
+  params_ = svd.matrixV().col(0);
+  intercept_(0) = -mean(0, 0);
+  intercept_(1) = -mean(0, 1);
+  intercept_(2) = -mean(0, 2);
+  params_.normalize();
+  double squared_error = distFromPoint(new_point);
+  for (Prototype pt : model_points_)
+  {
+    squared_error += distFromPoint(pt);
+  }
+  if (squared_error > error_t_)
+  {
+    params_ = old_params;
+    intercept_ = old_intercept;
+    return false;
+  }
+  model_points_.emplace_back(new_point);
+  end_point_ = new_point;
+  return true;
+}
+
+FastSegmentFilter::FastSegmentFilter(const ros::NodeHandle &nh) : private_nh_{ nh }, config_{ nh }
+{
+  ground_pub_ = private_nh_.advertise<pcl::PointCloud<velodyne_pointcloud::PointXYZIR>>(config_.ground_topic, 1);
+  nonground_pub_ = private_nh_.advertise<pcl::PointCloud<velodyne_pointcloud::PointXYZIR>>(config_.nonground_topic, 1);
+  marker_pub_ = private_nh_.advertise<visualization_msgs::MarkerArray>("Lines_array", 1);
+  ;
+}
+
+void FastSegmentFilter::processSegments()
+{
+  for (const auto &seg : segments_)
+  {
+    std::map<int, std::vector<velodyne_pointcloud::PointXYZIR>> bins_raw;
+    for (const auto &point : seg.second.raw_points_)
+    {
+      bins_raw[point.ring].emplace_back(point);
+    }
+    for (const auto &bin : bins_raw)
+    {
+      velodyne_pointcloud::PointXYZIR prototype_pt = bin.second[0];
+      for (const auto &point : bin.second)
+      {
+        if (point.z < prototype_pt.z)
+        {
+          prototype_pt = point;
+        }
+      }
+      Prototype ptype = { getDistanceFromPoint(prototype_pt), prototype_pt, 0, 0, 0 };
+      segments_[seg.first].prototype_points_.emplace_back(ptype);
+    }
+  }
+}
+
+void FastSegmentFilter::getLinesFromSegments()
+{
+  for (auto &seg : segments_)
+  {
+    std::vector<Prototype> sorted_points = seg.second.prototype_points_;
+    sort(sorted_points.begin(), sorted_points.end());
+    Line curr_line = Line(config_.error_t);
+    for (Prototype pt : sorted_points)
+    {
+      if (!curr_line.fitPoints(pt))
+      {
+        curr_line.isGround_ = evaluateIsGround(curr_line);
+        seg.second.lines_.emplace_back(curr_line);
+        curr_line = Line(config_.error_t);
+        curr_line.fitPoints(pt);
+      }
+    }
+    curr_line.isGround_ = evaluateIsGround(curr_line);
+    for (Prototype &pt : curr_line.model_points_)
+    {
+      pt.blue_ = curr_line.blue_;
+      pt.green_ = curr_line.green_;
+      pt.red_ = curr_line.red_;
+    }
+    seg.second.lines_.emplace_back(curr_line);
+  }
+}
+
+void FastSegmentFilter::classifyPoints(pcl::PointCloud<velodyne_pointcloud::PointXYZIR> &ground_points,
+                                       pcl::PointCloud<velodyne_pointcloud::PointXYZIR> &nonground_points)
+{
+  for (const auto &seg : segments_)
+  {
+    Segment segment = seg.second;
```

Comment:
Maybe do `for (const auto& [segment_id, segment] : segments_)` to make it more concise?